File: 852_Peak_Index_in_a_Mountain_Array/Solution.hs
{-
Leetcode 852. peak index in a mountain array
binary search for peak in strictly increasing then decreasing array
-}

-- Generic binary search: find first index where f i == True
binarySearch :: (Int -> Bool) -> Int -> Int -> Int
binarySearch f low high = go low high
  where
    go l r
      | l < r =
          let mid = (l + r) `div` 2
          in if f mid
               then go l mid
               else go (mid + 1) r
      | otherwise = l

peakIndexInMountainArray :: [Int] -> Int
peakIndexInMountainArray xs =
  binarySearch (\i -> xs !! i > xs !! (i + 1)) 0 (length xs - 1)

main :: IO ()
main = do
  print $ peakIndexInMountainArray [0, 1, 2, 3, 2, 1]     -- 3
  print $ peakIndexInMountainArray [0, 2, 4, 6, 5, 3, 1]  -- 3
  print $ peakIndexInMountainArray [1, 3, 5, 4, 2]        -- 2
  print $ peakIndexInMountainArray [0, 10, 5, 2]          -- 1
  print $ peakIndexInMountainArray [0, 1, 0]              -- 1

File: 852_Peak_Index_in_a_Mountain_Array/solution.cpp
// Leetcode 852. Peak Index in a Mountain Array
// Binary search for peak in strictly increasing then decreasing array
//
// 我們的目標是：
//   找到第一個 index，使得 arr[i] > arr[i + 1]，也就是「轉折點」。
//   So while (left < right)

#include <vector>

int peakIndexInMountainArray(std::vector<int>& arr) {
  int left = 0, right = arr.size() - 1;
  while (left < right) {
    int mid = left + (right - left) / 2;
    if (arr[mid] < arr[mid + 1]) {
      left = mid + 1; // peak is to the right
    } else {
      right = mid;    // peak is at mid or to the left
    }
  }
  return left; // left == right == peak index
}
